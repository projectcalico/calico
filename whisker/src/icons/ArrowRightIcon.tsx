import { createIcon } from '@chakra-ui/icons';

const CalicoCatIcon = createIcon({
    displayName: 'ArrowRightIcon',
    viewBox: '0 0 9 7',
    path: (
        <path d='M5.88672 0.515625L5.49609 0.90625C5.41797 1.00391 5.41797 1.14062 5.51562 1.23828L7.07812 2.74219H0.359375C0.222656 2.74219 0.125 2.85938 0.125 2.97656V3.52344C0.125 3.66016 0.222656 3.75781 0.359375 3.75781H7.07812L5.51562 5.28125C5.41797 5.37891 5.41797 5.51562 5.49609 5.61328L5.88672 6.00391C5.98438 6.08203 6.12109 6.08203 6.21875 6.00391L8.79688 3.42578C8.875 3.32812 8.875 3.19141 8.79688 3.09375L6.21875 0.515625C6.12109 0.4375 5.98438 0.4375 5.88672 0.515625Z' />
    ),
});

export default CalicoCatIcon;
