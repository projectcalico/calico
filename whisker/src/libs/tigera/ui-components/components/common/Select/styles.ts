export default {
    control: (provided: any) => ({
        ...provided,
        height: '42px',
        boxShadow: 'none',
        ':focus': {
            color: 'tigeraRed.800',
        },
        '&[data-invalid="true"]': { boxShadow: 'none' },
        bg: 'green',

        backgroundColor: 'green',
    }),
    container: (provided: any) => ({
        ...provided,
        width: 'full',
        backgroundColor: 'green',
    }),

    clearIndicator: (provided: any) => ({
        ...provided,
        fontSize: '0.60rem',
        color: 'tigeraGrey.700',
        '--close-button-size': '10px',
        ':hover': {
            bg: 'unset',
            color: 'tigeraGrey.900',
        },
    }),
    indicatorsContainer: (provided: any) => ({
        ...provided,
        fontSize: 'xl',
    }),
    multiValueRemove: (provided: any) => ({
        ...provided,
        fontSize: 'sm',
    }),
    menu: (provided: any) => ({
        ...provided,
        mt: 1,
        boxShadow: '0px 0px 8px #dcdde0 !important',
        borderColor: 'tigeraGrey.300',
        borderRadius: 'md',
        bg: 'green',
    }),
    menuList: (provided: any) => ({
        ...provided,
        pt: 0,
        pb: 0,
        bg: 'green',
    }),
    noOptionsMessage: (state: any) => ({
        ...state,
        textAlign: 'left',
        px: 4,
    }),
    option: (state: any) => ({
        ...state,
        fontFamily: 'Poppins',
        fontWeight: 400,
        px: 2.5,
        py: 2,
        fontSize: 'sm',
        minWidth: 'fit-content',
        ':hover': {
            bg: 'tigeraGrey.200',
            color: 'tigeraBlack',
        },
    }),
    dropdownIndicator: (provided: any) => ({
        ...provided,
        bg: 'transparent',
        fontSize: 'xl',
        px: 2,
        cursor: 'inherit',
    }),
    indicatorSeparator: (provided: any) => ({
        ...provided,
        display: 'none',
    }),
    singleValue: (provided: any) => ({
        ...provided,
    }),
    multiValue: (provided: any) => ({
        ...provided,
        py: 4,
        borderRadius: 0,
        borderWidth: 0,
        bg: '#ebf5ff', // TODO THIS COLOR IS NOT IN THE DESIGN THEME! NEEDS FEEDBACK FROM DESIGNERS
        svg: {
            color: 'white',
        },
        p: 0,
        pr: 1,
        fontSize: 'xs',
        lineHeight: 5,
        height: 6,
        '>span': {
            py: 3,
            px: '5px',
        },
        '>div': {
            bg: 'tigeraBlueDark',
            opacity: 1,
            borderRadius: 0,
            height: 6,
            width: 6,
        },
    }),
};
