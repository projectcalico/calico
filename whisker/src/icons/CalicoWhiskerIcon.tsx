import { createIcon } from '@chakra-ui/icons';

const CalicoCatIcon = createIcon({
    displayName: 'CalicoWhiskerIcon',
    viewBox: '0 0 14 19',
    path: (
        <>
            <path
                d='M12.3848 6.08052L12.3939 6.08496L12.4038 6.08746C12.4682 6.10387 12.5349 6.10417 12.5994 6.08837C12.7094 6.06801 12.8082 6.00717 12.8839 5.91889C12.9602 5.82984 13.0104 5.71626 13.0303 5.59467C13.0502 5.47303 13.0392 5.34655 12.9982 5.23194C12.9573 5.11739 12.8874 5.01772 12.7949 4.94922L12.7875 4.94367L12.7791 4.93961L4.61031 0.975028C4.54633 0.933769 4.47481 0.908356 4.40022 0.901732C4.32323 0.894895 4.24646 0.90839 4.17538 0.94048C4.10447 0.972496 4.04162 1.0219 3.99017 1.08385C3.93875 1.14576 3.89952 1.2192 3.87407 1.29885C3.84862 1.37852 3.83727 1.46341 3.84047 1.54813C3.84366 1.63284 3.86136 1.7163 3.89287 1.79313C3.92437 1.86995 3.96931 1.93917 4.02575 1.9955C4.08049 2.05013 4.14508 2.09164 4.21606 2.11597L12.3848 6.08052Z'
                fill='#F69320'
                stroke='#F69320'
                stroke-width='0.2'
            />
            <path
                d='M1.87473 10.3346L1.87473 10.3345L12.4281 10.2037C12.4284 10.2037 12.4286 10.2037 12.4288 10.2037C12.5934 10.2035 12.7496 10.1329 12.8635 10.0101C12.9774 9.88736 13.0402 9.72242 13.0402 9.55185C13.0402 9.38129 12.9774 9.21634 12.8635 9.09361C12.7494 8.97064 12.593 8.9 12.4281 8.89999L12.4268 8.90001L1.8734 9.04031L1.87339 9.0395L1.86074 9.04128C1.7114 9.06237 1.57595 9.14106 1.47883 9.26035C1.38184 9.37946 1.3292 9.53132 1.3292 9.68747C1.3292 9.84363 1.38184 9.99548 1.47883 10.1146C1.57595 10.2339 1.7114 10.3126 1.86074 10.3337L1.86582 10.2977L1.86605 10.3346H1.87473Z'
                fill='#F69320'
                stroke='#F69320'
                stroke-width='0.2'
            />
            <path
                d='M12.3848 13.1541L12.3939 13.1497L12.4038 13.1472C12.4682 13.1308 12.5349 13.1305 12.5994 13.1463C12.7094 13.1666 12.8082 13.2275 12.8839 13.3158C12.9602 13.4048 13.0104 13.5184 13.0303 13.64C13.0502 13.7616 13.0392 13.8881 12.9982 14.0027C12.9573 14.1173 12.8874 14.2169 12.7949 14.2854L12.7875 14.291L12.7791 14.295L4.61031 18.2596C4.54633 18.3009 4.47481 18.3263 4.40022 18.3329C4.32323 18.3398 4.24646 18.3263 4.17538 18.2942C4.10447 18.2622 4.04162 18.2128 3.99017 18.1508C3.93875 18.0889 3.89952 18.0154 3.87407 17.9358C3.84862 17.8561 3.83727 17.7712 3.84047 17.6865C3.84366 17.6018 3.86136 17.5183 3.89287 17.4415C3.92437 17.3647 3.96931 17.2955 4.02575 17.2391C4.08049 17.1845 4.14508 17.143 4.21606 17.1187L12.3848 13.1541Z'
                fill='#F69320'
                stroke='#F69320'
                stroke-width='0.2'
            />
        </>
    ),
});

export default CalicoCatIcon;
