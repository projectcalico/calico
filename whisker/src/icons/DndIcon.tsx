import { createIcon } from '@chakra-ui/icons';

const DndIcon = createIcon({
    displayName: 'DndIcon',
    viewBox: '0 0 16 8',
    path: (
        <path
            d='M15.0312 0.90625H0.937109C0.473047 0.90625 0.125 1.28594 0.125 1.71836C0.125004 2.21406 0.504688 2.59375 0.937109 2.59375H15.0312C15.4953 2.59375 15.875 2.21406 15.875 1.78164C15.875 1.28594 15.4953 0.90625 15.0312 0.90625ZM15.0312 5.40625H0.937109C0.473047 5.40625 0.125 5.78594 0.125 6.21836C0.125 6.65078 0.504688 7.09375 0.937109 7.09375H15.0312C15.4953 7.09375 15.875 6.71406 15.875 6.28164C15.875 5.84922 15.4953 5.40625 15.0312 5.40625Z'
            fill='#5F5F5F'
        />
    ),
});

export default DndIcon;
