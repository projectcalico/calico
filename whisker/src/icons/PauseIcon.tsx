import { createIcon } from '@chakra-ui/icons';

const PauseIcon = createIcon({
    displayName: 'PauseIcon',
    viewBox: '0 0 320 512',
    path: (
        <path d='M48 64C21.5 64 0 85.5 0 112L0 400c0 26.5 21.5 48 48 48l32 0c26.5 0 48-21.5 48-48l0-288c0-26.5-21.5-48-48-48L48 64zm192 0c-26.5 0-48 21.5-48 48l0 288c0 26.5 21.5 48 48 48l32 0c26.5 0 48-21.5 48-48l0-288c0-26.5-21.5-48-48-48l-32 0z' />
    ),
});

export default PauseIcon;
