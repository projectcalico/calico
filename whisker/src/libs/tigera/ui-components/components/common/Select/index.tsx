import * as React from 'react';
import * as ReactSelect from 'chakra-react-select';
import styles from './styles';
import type { GroupBase, Props } from 'react-select';
import { Box, Flex, Icon, Image, forwardRef } from '@chakra-ui/react';

type Option = any;
type IsMulti = boolean;
type Group = GroupBase<Option>;
type ChakraStylesConfig = {
    sx?: ReactSelect.ChakraStylesConfig;
};
export type SelectType = Props<Option, IsMulti, Group> & ChakraStylesConfig;
export type CreatableSelectType = ReactSelect.CreatableProps<
    Option,
    IsMulti,
    Group
> &
    ChakraStylesConfig;

export type SelectOption = {
    label: string;
    value: string;
    iconUrl?: string;
};

// TODO this will require more tweaking depending on specific use cases - change it as required
// its not styled in the same way as a standard Chakra component...
// see here for more: https://github.com/csandman/chakra-react-select#readme

const Select: React.FC<React.PropsWithChildren<SelectType>> = forwardRef(
    ({ sx, isMulti = false, options, ...rest }, ref) => {
        const hasIcons = options?.every(
            (option: Option) =>
                option.icon !== undefined || option.iconUrl !== undefined,
        );

        return (
            <ReactSelect.Select
                ref={ref}
                isMulti={isMulti}
                options={options}
                {...((hasIcons
                    ? {
                          formatOptionLabel: (e: any) => (
                              <SelectIconOption
                                  icon={e.icon}
                                  iconUrl={e.iconUrl}
                                  label={e.label}
                              />
                          ),
                      }
                    : {}) as any)}
                {...rest}
                chakraStyles={{ ...styles, ...sx }}
            />
        );
    },
);

export const CreatableSelect: React.FC<CreatableSelectType> = ({
    sx,
    ...rest
}) => (
    <ReactSelect.CreatableSelect
        {...rest}
        chakraStyles={{ ...styles, ...sx }}
    />
);

export const SelectIconOption: React.FC<
    React.PropsWithChildren<{
        label: string;
        icon?: any;
        iconUrl?: string;
    }>
> = ({ icon, iconUrl, label }) => (
    <Flex alignItems='center'>
        {icon && <Icon as={icon} w={6} h={6} mr={2} />}
        {iconUrl && <Image src={iconUrl} w={6} h={6} />}
        <Box as='span' ml={2}>
            {label}
        </Box>
    </Flex>
);

export default Select;
