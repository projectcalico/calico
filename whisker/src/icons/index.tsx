import CalicoCatIcon from './CalicoCatIcon';
import TableSortIcon from './TableSortIcon';
import TableAscendingSortIcon from './TableAscendingSortIcon';
import CalicoWhiskerIcon from './CalicoWhiskerIcon';
import ArrowRightIcon from './ArrowRightIcon';

export {
    CalicoCatIcon,
    TableSortIcon,
    TableAscendingSortIcon,
    CalicoWhiskerIcon,
    ArrowRightIcon,
};
