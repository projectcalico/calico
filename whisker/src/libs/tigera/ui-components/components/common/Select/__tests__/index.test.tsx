import { render, screen } from '@/test-utils/helper';
import { ArrowDownIcon, ArrowUpIcon } from '@chakra-ui/icons';
import { Box } from '@chakra-ui/react';
import { fireEvent } from '@testing-library/react';
import React from 'react';
import Select, { CreatableSelect, SelectIconOption } from '../index';

jest.mock('chakra-react-select', () => ({
    Select: ({ options = [], onChange, isMulti, ...rest }: any) => {
        const [values, setValues] = React.useState<any[]>([]);
        const handleChange = (value: any) => {
            const option = options.find(
                (option: { value: string }) => option.value === value,
            );
            onChange(option);
        };

        const handleMultiChange = (value: any) => {
            const option = options.find(
                (option: { value: string }) => option.value === value,
            );
            const newValues = [...values, option];
            setValues(newValues);
            onChange(newValues);
        };

        return (
            <select data-testid='react-select' {...rest}>
                {options.map(({ label, value }: any) => (
                    <option
                        key={value}
                        value={value}
                        onClick={() =>
                            isMulti
                                ? handleMultiChange(value)
                                : handleChange(value)
                        }
                    >
                        {label}
                    </option>
                ))}
            </select>
        );
    },
    CreatableSelect: (props: any) => <Box {...props}>MockCreatableSelect</Box>,
}));

export const mockOptions = [
    { value: 'blue', label: 'Blue', color: '#0052CC' },
    { value: 'purple', label: 'Purple', color: '#5243AA' },
];

export const mockOptionsWithIcons = [
    {
        value: 'blue',
        label: 'Blue',
        color: '#0052CC',
        icon: ArrowUpIcon,
    },
    {
        value: 'purple',
        label: 'Purple',
        color: '#5243AA',
        icon: ArrowDownIcon,
    },
];

describe('<Select/>', () => {
    it('it renders with options', () => {
        const mockedOnChange = jest.fn();
        const { queryByTestId } = render(
            <div data-testid='my-select-component'>
                <Select
                    options={mockOptionsWithIcons}
                    onChange={mockedOnChange}
                />
            </div>,
        );

        const mySelectComponent = queryByTestId(
            'my-select-component',
        ) as HTMLElement;

        expect(mySelectComponent).toBeDefined();
        expect(mySelectComponent).not.toBeNull();
        expect(mockedOnChange).toHaveBeenCalledTimes(0);

        fireEvent.click(screen.getByTestId('react-select'));
        fireEvent.click(screen.getByText('Blue'));

        expect(mockedOnChange).toHaveBeenCalledTimes(1);
    });

    it('it renders with no options', () => {
        const { asFragment } = render(
            <div data-testid='my-select-component'>
                <Select />
            </div>,
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it('it renders SelectIconOption with icon', () => {
        const { asFragment } = render(
            <div data-testid='my-select-component'>
                <SelectIconOption label={'test'} icon={ArrowUpIcon} />
            </div>,
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it('it renders SelectIconOption with iconUrl', () => {
        const { asFragment } = render(
            <div data-testid='my-select-component'>
                <SelectIconOption
                    label={'test'}
                    iconUrl='https://www.mockurl.com/image.png'
                />
            </div>,
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it('it renders with no options', () => {
        const { asFragment } = render(
            <div data-testid='my-select-component'>
                <Select
                    id='form-field'
                    name='colors'
                    selectedOptionStyle='check'
                    options={mockOptions}
                    isMulti={true}
                    placeholder='Select...'
                    closeMenuOnSelect={false}
                />
            </div>,
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it('it renders creatable', () => {
        const { asFragment } = render(
            <div data-testid='my-select-component'>
                <CreatableSelect
                    id='form-field'
                    name='colors'
                    selectedOptionStyle='check'
                    options={mockOptions}
                    isMulti={true as any}
                    placeholder='Select...'
                    closeMenuOnSelect={false}
                />
                ,
            </div>,
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it('it can select multiple values', () => {
        const mockOnChange = jest.fn();
        const { asFragment } = render(
            <div data-testid='my-select-component'>
                <Select
                    id='form-field'
                    name='colors'
                    selectedOptionStyle='check'
                    options={mockOptions}
                    isMulti={true as any}
                    placeholder='Select...'
                    closeMenuOnSelect={false}
                    classNamePrefix='multi'
                    onChange={mockOnChange}
                />
                ,
            </div>,
        );

        fireEvent.keyDown(
            screen.getByTestId('my-select-component').firstChild as any,
            {
                key: 'ArrowDown',
            },
        );
        fireEvent.click(screen.getByText('Blue'));
        fireEvent.keyDown(
            screen.getByTestId('my-select-component').firstChild as any,
            {
                key: 'ArrowDown',
            },
        );
        fireEvent.click(screen.getByText('Purple'));

        expect(mockOnChange).toHaveBeenCalledWith(mockOptions);
        expect(asFragment()).toMatchSnapshot();
    });
});
