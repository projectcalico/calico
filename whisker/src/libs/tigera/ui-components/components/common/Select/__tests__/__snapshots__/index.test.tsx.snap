// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select/> it can select multiple values 1`] = `
<DocumentFragment>
  <div
    data-testid="my-select-component"
  >
    <select
      chakrastyles="[object Object]"
      classnameprefix="multi"
      data-testid="react-select"
      id="form-field"
      name="colors"
      placeholder="Select..."
      selectedoptionstyle="check"
    >
      <option
        value="blue"
      >
        Blue
      </option>
      <option
        value="purple"
      >
        Purple
      </option>
    </select>
    ,
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<Select/> it renders SelectIconOption with icon 1`] = `
<DocumentFragment>
  <div
    data-testid="my-select-component"
  >
    <div
      class="css-70qvj9"
    >
      <svg
        class="chakra-icon chakra-icon css-12m389b"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M4 12l1.41 1.41L11 7.83V20h2V7.83l5.58 5.59L20 12l-8-8-8 8z"
          fill="currentColor"
        />
      </svg>
      <span
        class="css-1v4xcoh"
      >
        test
      </span>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<Select/> it renders SelectIconOption with iconUrl 1`] = `
<DocumentFragment>
  <div
    data-testid="my-select-component"
  >
    <div
      class="css-70qvj9"
    >
      <img
        class="chakra-image css-4n3cc1"
        src="https://www.mockurl.com/image.png"
      />
      <span
        class="css-1v4xcoh"
      >
        test
      </span>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<Select/> it renders creatable 1`] = `
<DocumentFragment>
  <div
    data-testid="my-select-component"
  >
    <div
      chakrastyles="[object Object]"
      class="css-0"
      id="form-field"
      name="colors"
      options="[object Object],[object Object]"
      placeholder="Select..."
      selectedoptionstyle="check"
    >
      MockCreatableSelect
    </div>
    ,
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<Select/> it renders with no options 1`] = `
<DocumentFragment>
  <div
    data-testid="my-select-component"
  >
    <select
      chakrastyles="[object Object]"
      data-testid="react-select"
    />
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<Select/> it renders with no options 2`] = `
<DocumentFragment>
  <div
    data-testid="my-select-component"
  >
    <select
      chakrastyles="[object Object]"
      data-testid="react-select"
      id="form-field"
      name="colors"
      placeholder="Select..."
      selectedoptionstyle="check"
    >
      <option
        value="blue"
      >
        Blue
      </option>
      <option
        value="purple"
      >
        Purple
      </option>
    </select>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
